src/main/java/com/solace/samples/java/patterns/GuaranteedProcessor.java

```diff
@@ -0,0 +1,227 @@
+    
+    private static volatile int msgSentCounter = 0;                 // num messages sent
+    private static volatile int msgRecvCounter = 0;                 // num messages received
+    private static volatile boolean hasDetectedRedelivery = false;  // detected any messages being redelivered?
+    private static volatile boolean isShutdown = false;             // are we done?
+
+    // remember to add log4j2.xml to your classpath
+    private static final Logger logger = LogManager.getLogger();  // log4j2, but could also use SLF4J, JCL, etc.
+
+    /** This is the main app.  Use this type of app for receiving Guaranteed messages (e.g. via a queue endpoint). */
+    public static void main(String... args) throws InterruptedException, IOException {
+        if (args.length < 3) {  // Check command line arguments
+            System.out.printf("Usage: %s <host:port> <message-vpn> <client-username> [password]%n%n", SAMPLE_NAME);
+            System.exit(-1);
+        }
+        System.out.println(API + " " + SAMPLE_NAME + " initializing...");
+
+        final Properties properties = new Properties();
+        properties.setProperty(TransportLayerProperties.HOST, args[0]);          // host:port
+        properties.setProperty(ServiceProperties.VPN_NAME,  args[1]);     // message-vpn
+        properties.setProperty(AuthenticationProperties.SCHEME_BASIC_USER_NAME, args[2]);      // client-username
+        if (args.length > 3) {
+            properties.setProperty(AuthenticationProperties.SCHEME_BASIC_PASSWORD, args[3]);  // client-password
+        }
+        //properties.setProperty(JCSMPProperties.GENERATE_SEQUENCE_NUMBERS, true);  // not required, but interesting
+        
+        properties.setProperty(TransportLayerProperties.RECONNECTION_ATTEMPTS, "20");  // recommended settings
+        properties.setProperty(TransportLayerProperties.CONNECTION_RETRIES_PER_HOST, "5");
+        // https://docs.solace.com/Solace-PubSub-Messaging-APIs/API-Developer-Guide/Configuring-Connection-T.htm
+        // ready to connect now
+        final MessagingService messagingService = MessagingService.builder(ConfigurationProfile.V1)
+                .fromProperties(properties)
+                .build();
+        messagingService.connect();  // blocking connect
+        messagingService.addServiceInterruptionListener(serviceEvent -> {
+            logger.info("### SERVICE INTERRUPTION: "+serviceEvent.getCause());
+            //isShutdown = true;
+        });
+        messagingService.addReconnectionAttemptListener(serviceEvent -> {
+            logger.info("### RECONNECTING ATTEMPT: "+serviceEvent);
+        });
+        messagingService.addReconnectionListener(serviceEvent -> {
+            logger.info("### RECONNECTED: "+serviceEvent);
+        });
+
+        
+        // build the publisher object
+        final PersistentMessagePublisher publisher = messagingService.createPersistentMessagePublisherBuilder()
+                .onBackPressureWait(1)
```

Comment:
I def think it's best practice, and wish it was the default rather than elastic, which will use up all your heap if you're offline for too long.  I think they're changing it in v1.1..?